src/utils/page-functions.js

```diff
@@ -0,0 +1,202 @@
+import { createHash } from 'crypto'
+import path from 'path'
+import webpack from 'webpack'
+import { camelCase } from '../utils'
+import {
+  readFile,
+  exists,
+  stats,
+  requireResolve
+} from './fs'
+
+let packageName
+let nodeModulesPath
+
+/* export async function listPageFunctionSources(fnPath) {
+  if (!fnPath) {
+    fnPath = path.join(__dirname, 'page-functions')
+  }
+
+  const files = await readDir(fnPath, { encoding: 'utf8' })
+  return files.filter(file => file.endsWith('.js'))
+} */
+
+export function createPageFunctions(page, sourceFiles) {
+  const pageFunctions = {}
+
+  if (!sourceFiles) {
+    return pageFunctions
+  }
+
+  const babelPresets = page.getBabelPresetOptions()
+
+  for (let file of sourceFiles) {
+    let fnName
+    if (Array.isArray(file)) {
+      [file, fnName] = file
+    } else {
+      fnName = camelCase(path.basename(file, '.js'))
+    }
+
+    pageFunctions[fnName] = async (...args) => {
+      const body = await getPageFunctionBody(fnName, file, babelPresets)
+
+      return page.runAsyncScript({
+        args: [],
+        body: `${body} return (pageFn.${fnName} || pageFn.default).apply(null, arguments)`
+      }, ...args)
+    }
+  }
+
+  return pageFunctions
+}
+
+export async function getPageFunctionBody(fnName, filePath, presetOptions = {}) {
+  const cacheKey = createCacheKey(filePath, presetOptions)
+  const cacheFile = `${fnName}-${cacheKey}.js`
+  const cachePath = await getCacheEntry(cacheFile)
+  let cacheValid = await exists(cachePath)
+
+  if (cacheValid) {
+    const { mtime: fm } = await stats(filePath)
+    const { mtime: cm } = await stats(cachePath)
+
+    if (fm >= cm) {
+      cacheValid = false
+    }
+  }
+
+  if (!cacheValid) {
+    // transpile page function
+    await compilePageFunction({
+      name: 'pageFn',
+      filePath,
+      cachePath,
+      presetOptions
+    })
+  }
+
+  const fnBody = await readFile(cachePath, { encoding: 'utf8' })
+  return fnBody
+}
+
+export function compilePageFunction(config) {
+  return new Promise((resolve, reject) => {
+    const webpackConfig = createWebpackConfig(config)
+
+    webpack(webpackConfig, (err, stats) => {
+      /* istanbul ignore next */
+      if (err) {
+        reject(err)
+        return
+      }
+
+      if (stats.hasErrors()) {
+        const [err] = stats.compilation.errors
+        reject(err)
+        return
+      }
+
+      resolve(stats)
+    })
+  })
+}
+export function createWebpackConfig({
+  name,
+  filePath,
+  cachePath,
+  presetOptions
+}) {
+  return {
+    mode: 'production',
```

Comment:
Not sure but when using async chunks, weback may split output to serval files. Causing problems when loading body. We can use this plugin with limit of 1: https://webpack.js.org/plugins/limit-chunk-count-plugin/ in that case

Reply:
Thanks, will add this